doc/design/api.md

```diff
@@ -0,0 +1,90 @@
+# Design Doc: PaddlePaddle API
+
+## Ingredients
+
+As the first step of our design, we list important concepts in deep
+learning and try to figure their relationship, as shown below:
+
+```
+Model = {topology, parameters}
+
+Evaluator = {Model*, activations}
+- forward
+- test
+
+GradientMachine = {Model*, gradients}
+- backward
+
+Optimizer = {Model*, Evaluator*, GradientMachine*}
+- train
+- update
```

Comment:
If we are going to use *composition* other than *derivation*, then I would agree that we should have Optimizer, which has the `train` method, and Updater, which has the `update` method.  But as Python supports derivation, we can create AdamOptimizer basing on SGDOptimizer by simply overriding `Optimizer::update`.